runtime/compiler/optimizer/J9TransformUtil.hpp

```diff
@@ -201,8 +201,23 @@ class OMR_EXTENSIBLE TransformUtil : public OMR::TransformUtilConnector
    static bool fieldShouldBeCompressed(TR::Node *node, TR::Compilation *comp);
 
    static TR::Block *insertNewFirstBlockForCompilation(TR::Compilation *comp);
-   static TR::Node *calculateOffsetFromIndexInContiguousArray(TR::Compilation *, TR::Node * index, TR::DataType type);
-   static TR::Node *calculateElementAddress(TR::Compilation *, TR::Node *array, TR::Node * index, TR::DataType type);
+   /**
+    * \brief
+    *    Calculate offset for an array element given its index based on `type` or `elementStride`
+    *
+    * \param index
+    *    The index into the array
+    *
+    * \param type
+    *    The data type of the array element. If `elementStride` is none zero, `type` is ignored
+    *
+    * \param elementStride
+    *    The size of the array element. If it is zero, the width is calculated based on TR::DataType `type`
+    *
+    * \return The Int64 value representing the offset into an array of the specified type given the index
+    */
+   static TR::Node *calculateOffsetFromIndexInContiguousArray(TR::Compilation *, TR::Node * index, TR::DataType type, int32_t elementStride = 0);
```

Comment:
Zero might be a valid element stride for an array of a value type that has no fields.  I'm just verifying that.  If I'm correct, you'll need to use a different way of signalling that the `type` parameter should be used.

One possibility might be to split this into two methods - one that accepts only the `type` parameter, and one that accepts only the `elementStride` parameter, and then the version that uses the `type` parameter could call the version that uses `elementStride`.

Another possibility might be to use a negative stride.

Reply:
Thanks for the suggestion! I'll split them into two APIs:

One takes `TR::DataType type`
```
   static TR::Node *calculateOffsetFromIndexInContiguousArray(TR::Compilation *, TR::Node * index, TR::DataType type);
   static TR::Node *calculateElementAddress(TR::Compilation *, TR::Node *array, TR::Node * index, TR::DataType type);
```
Another takes `int32_t elementStride`
```
   static TR::Node *calculateOffsetFromIndexInContiguousArrayWithElementStride(TR::Compilation *, TR::Node * index, int32_t elementStride);
   static TR::Node *calculateElementAddressWithElementStride(TR::Compilation *, TR::Node *array, TR::Node * index, int32_t elementStride);
```

Reply:
Addressed in cf9ae54

Reply:
I noticed an issue with the refactored code in 7bfe3e8. `_widthToShift[]` has size 9 but `elementStride` could be any value

```
   if (elementStride && (elementStride % 2 == 0))
       shift = TR::TransformUtil::convertWidthToShift(elementStride);
``

Reply:
I have a fix now. Change the status to `WIP` while I'm running more tests.